### DISCUSSION TITLE
Zoom factor, how to use

### DISCUSSION THREAD

**Discussion by akraskovSMDB** [neuropoly/axondeepseg#732] [Help needed]
Hi, 

I am trying to understand how to use a new zoom factor option for segmentation and whether this option can help me to segment smaller axons.

My segmentation (based on my own model) is missing quite a bit of small axons (~<1um). When I looked into the training set it also missed some small axons. Therefore, I think, my model is not great in detecting them on unseen images. 

The obvious way to correct this would be to go back to the training set and painstakingly add more small axons. BUT, zoom option seems to suggest that there is a simple shortcut. Assuming that, e.g. axons of 2um size range (say 1.5-2.5um) are similar (in some general sense) to axons with 1um size range, wouldn't be possible to use model knowledge about 2um range axons to segment 1um range axons (i.e. 0.75-1.25um)?

To try this out, I first need to understand whether I have to use zoom factor 2 or 0.5?
Secondly, if my model is trained using 256x256 patches, will the zoom option work if my segmentation is done with --no-patch option? Or I should not use --no-patch option and add overlap option instead?

Best wishes
Alexander

**Comment by Stoyan-I-A** [comment]
The zoom factor will essentially rescale your image by this factor. So with a zoom factor of 2, your 1um axons will look like 2um axons, but your previously 2um axons will look like 4um axons. 
It does not affect the size of the patches. If your image has dimensions of 1000 x 1000 and your model uses 256 x 256 patches, a zoom factor of 2 will result in having a 2000 x 2000 image that will be segmented using more 256 x 256 patches than with a factor of 1.

You could try to segment an image with a zoom factor of 2 and compare the results with the default zoom factor of 1. And perhaps you could combine elements of the segmentation using an image editing software. 
You could also try to use the [zoom factor sweep](https://axondeepseg.readthedocs.io/en/latest/documentation.html#segment-an-image-using-a-range-of-zoom-factors) to test out multiple zoom factors with one command.

@hermancollin Could you confirm that what I say is right?

**Comment by akraskovSMDB** [comment]
Thank you, @Stoyan-I-A 

Just to make sure, with zoom factor 2, an initial (say 1k x 1k) image will be resampled to a larger image (2k x 2k).

Is there any other resampling during the segmentation stage?  Similar to resampling happening during model training governed by the hspace and wspace parameters?

**Reply by Stoyan-I-A** [reply]
The ratio of (`pixel size of your image`) : (`pixel size used to train the model`) is also used to resample the image but this is done at the same time as the zoom factor.
I don't think there's another resampling in the pipeline.

@hermancollin Can you confirm?

**Reply by mathieuboudreau** [reply]
Hi @akraskovSMDB ! Just curious as to why you'd like this information? Is it because you'd like to report the effective image pixel "size" that was used during the inference step in a paper/report? If so, once you've optimized the zoom factor that works best for you, we could certainly go look in the code and calculate that for you, or even open an issue so this value could be reported to the user.

During the morphometry calculations, values will always be calculated correctly if you've inputted your original image's pixel resolution.

The purpose of the "zoom" factor is that, although we've trained our models to certain "pixel resolutions", the machine learning "black box" doesn't know what these values are (they are never used in the training stage), the only thing it really "knows" (in my view) is the ratio between a pixel size and the typical axon sizes it "sees" (that we've provided). So, it's better optimized at inferring axon and myelin segmentations when this "ratio" is closer to what was used in the training, such that if your microscope zoom was substantially different than for our model, or your specimen axon sizes are also very different, we can use the zoom factor to bring the image patches closer to what the model was used to "seeing", and then it'll identify the axon and myelin regions better. Before the zoom factor feature, we used to "hack" this behaviour by changing the input pixel resolution directly, because we always resample the image back to the model's resolution (relative to the information given as the input resolution), but this would subsequently lead to incorrect morphometric values so we fixed this by writing the separate zoom factor implementation.

**Reply by hermancollin** [reply]
@akraskovSMDB sorry for the late reply, I missed this one.

To clarify, there is indeed a resampling step controlled by the `hspace` and `wspace` parameters in the config file, like you said. The zoom feature does not add a preliminary resampling before this, it only adjusts this resampling "rate".

**Reply by akraskovSMDB** [reply]
thank you @hermancollin and @mathieuboudreau.
I played a bit with zoom feature and it does not seem to help in my case. Please see attached my results for one [image](https://github.com/axondeepseg/axondeepseg/files/11490459/LM1_zooms_.pdf) 
Larger zooms (1.5, 2) seem to be missing inner axon space and split myelin for large axons. Overall, zoom 1 (no zoom) seems to give the best correspondence with manual segmentation.

Here is a summary plot of dice scores for 7 images across zooms
![dice](https://github.com/axondeepseg/axondeepseg/assets/25688017/ddfda2ff-97aa-4cf9-b5f4-aec1564603c9)

Do you have a similar experience? If not, do you see anything wrong with how segmentation results change with zoom?

**Reply by hermancollin** [reply]
@akraskovSMDB To me, it does make a lot of sense that zoom factor of 1 would give the best results for you, because IIUC, you trained your own model on your own data. This means that what the model saw during training matches exactly what you are trying to segment here. 

The original use-case of the zoom factor feature was to rescale user images to something similar to the data that the model saw during training. It has proven very useful for us multiple times when a researcher came up with images acquired at a resolution slightly different than the training data. In this scenario, the zoom factor was useful to adjust the resolution. In your case, this adjustment is unnecessary, and as you can see, it even hurts the performance of the model because it has the opposite effect: by changing the zoom factor, your target image does not match the training data anymore.

Hope this makes sense. However, I will say that at a quick glance, your segmentation with zoom factor 1 looks pretty good, doesn't it? A Dice score >95% is pretty impressive for an image with so many axons of different sizes. Is there something in particular that is bothering you with your results?

**Reply by mathieuboudreau** [reply]
Agreed - the segmentations a Zoom factor 1 look as good as we would expect with our models, some minor manual corrections are typically needed afterwards if desired but these look great!
